src/compiler/checker.ts

```diff
@@ -1013,6 +1013,12 @@ namespace ts {
             }
         }
 
+        function errorSkippedOn(key: keyof CompilerOptions, location: Node | undefined, message: DiagnosticMessage, arg0?: string | number, arg1?: string | number, arg2?: string | number, arg3?: string | number): Diagnostic {

```

Comment:
I'm having trouble wrapping my head around what this does and would find an explanation helpful.

Reply:
`location, message, arg0, ...` are the existing error logging parameters.

`key: keyof CompilerOptions` is a typesafe way to assert that `key`, while a string, actually refers to a property in `CompilerOptions`. That way the string can safely be used as a property name of `CompilerOptions` in `filterSemanticDiagnostics`:

`filter(diagnostics, d => !d.skippedOn || !options[d.skippedOn])`

This keeps diagnostics that 

1. were not issued by `errorSkippedOn`, so didn't set `skippedOn`
 -OR-
2. do have `skippedOn`, which is guaranteed to be the name of a property in compiler options, say "noEmit". For the "noEmit" example, `!options[d.skippedOn]` is equivalent to `!options.noEmit`.

This meta-programming JS would require an enum plus a switch-case conversion in C# I think. Or the meta-programming facilities there, though the result wouldn't be as typesafe as in Typescript.